File: backend/services/analyzer.py
"""
Deep Dive Analyzer Service - Fundamental and technical analysis

MCP-Ready Functions:
- analyze(): Full deep dive analysis
- check_solvency(): Operating cash flow check
- assess_volatility(): Volatility categorization
- assess_risk(): Overall risk level determination

Filters raw scan results to 3 viable candidates.
"""

from datetime import datetime
from typing import Optional, Tuple
from sqlmodel import Session

from backend.database import engine
from backend.models import AnalysisResult
from backend.services import yfinance_service, alpha_vantage
from backend.schemas.analysis import (
    AnalyzeRequest, AnalysisResponse,
    SolvencyCheck, VolatilityAssessment,
    RiskLevel, DirectionBias
)


def analyze(request: AnalyzeRequest) -> AnalysisResponse:
    """
    Perform full deep dive analysis on a ticker.

    MCP Tool: deep_dive_analysis

    Steps:
    1. Get current quote and price data
    2. Check solvency (Operating Cash Flow)
    3. Calculate 20-day volatility
    4. Get analyst ratings
    5. Assess risk level and direction bias
    6. Persist to database
    7. Return structured response
    """
    symbol = request.symbol.upper()

    # Step 1: Get quote
    quote = yfinance_service.get_quote(symbol)
    current_price = quote.get("price", 0)

    if not current_price:
        raise ValueError(f"Could not get price for {symbol}")

    # Step 2: Solvency check
    solvency = check_solvency(symbol)

    # Step 3: Volatility
    volatility = assess_volatility(symbol)

    # Step 4: Analyst ratings
    ratings = yfinance_service.get_analyst_ratings(symbol)

    # Step 5: Risk assessment
    risk_level, direction_bias, is_safe = assess_risk(
        volatility=volatility,
        solvency=solvency,
        price_vs_target=(ratings.get("upside_potential_pct") or 0),
        quote=quote
    )

    # Build recommendation
    recommendation = _build_recommendation(
        symbol, risk_level, direction_bias, is_safe, volatility, solvency
    )

    # Step 6: Persist
    with Session(engine) as session:
        analysis = AnalysisResult(
            scan_id=request.scan_id,
            symbol=symbol,
            current_price=current_price,
            price_change_pct=quote.get("change_percent"),
            fifty_two_week_high=quote.get("fifty_two_week_high"),
            fifty_two_week_low=quote.get("fifty_two_week_low"),
            operating_cash_flow=solvency.operating_cash_flow,
            is_solvent=solvency.is_solvent,
            volatility_20d=volatility.volatility_annualized,
            volatility_category=volatility.category.value,
            analyst_rating=ratings.get("recommendation_key"),
            analyst_target_mean=ratings.get("target_mean"),
            upside_potential_pct=ratings.get("upside_potential_pct"),
            risk_level=risk_level.value,
            direction_bias=direction_bias.value,
            is_safe_play=is_safe,
            raw_fundamentals=ratings,
            raw_technicals=quote
        )
        session.add(analysis)
        session.commit()
        session.refresh(analysis)
        analysis_id = analysis.id

    # Step 7: Return response
    return AnalysisResponse(
        analysis_id=analysis_id,
        symbol=symbol,
        current_price=current_price,
        solvency=solvency,
        volatility=volatility,
        analyst_rating=ratings.get("recommendation_key"),
        analyst_target_mean=ratings.get("target_mean"),
        upside_potential_pct=ratings.get("upside_potential_pct"),
        risk_level=risk_level,
        direction_bias=direction_bias,
        is_safe_play=is_safe,
        recommendation_summary=recommendation
    )


def check_solvency(symbol: str) -> SolvencyCheck:
    """
    Check solvency via Operating Cash Flow.

    MCP Tool: check_solvency

    Uses Alpha Vantage CASH_FLOW endpoint.
    Positive OCF = solvent.
    """
    try:
        cf_data = alpha_vantage.get_cash_flow(symbol)

        if "error" in cf_data:
            return SolvencyCheck(
                operating_cash_flow=None,
                free_cash_flow=None,
                is_solvent=True,
                notes="Unable to verify OCF - assuming solvent (verify manually)"
            )

        annual_reports = cf_data.get("annualReports", [])

        if annual_reports:
            latest = annual_reports[0]
            ocf_str = latest.get("operatingCashflow", "0")
            capex_str = latest.get("capitalExpenditures", "0")

            ocf = float(ocf_str) if ocf_str and ocf_str != "None" else 0
            capex = float(capex_str) if capex_str and capex_str != "None" else 0
            fcf = ocf - abs(capex)

            return SolvencyCheck(
                operating_cash_flow=ocf,
                free_cash_flow=fcf,
                is_solvent=ocf > 0,
                notes="Positive OCF indicates operational solvency" if ocf > 0 else "Negative OCF - cash burn"
            )
    except Exception as e:
        pass

    return SolvencyCheck(
        operating_cash_flow=None,
        free_cash_flow=None,
        is_solvent=True,
        notes="Unable to verify OCF - assuming solvent (verify manually)"
    )


def assess_volatility(symbol: str) -> VolatilityAssessment:
    """
    Calculate and categorize 20-day volatility.

    MCP Tool: assess_volatility
    """
    vol_data = yfinance_service.calculate_volatility(symbol, days=20)

    if "error" in vol_data:
        return VolatilityAssessment(
            volatility_20d=30.0,
            volatility_annualized=30.0,
            category=RiskLevel.MEDIUM,
            recommended_strategies=["Unable to calculate - use caution"]
        )

    category = RiskLevel(vol_data["category"])

    return VolatilityAssessment(
        volatility_20d=vol_data["daily_volatility"],
        volatility_annualized=vol_data["annualized_volatility"],
        category=category,
        recommended_strategies=vol_data["recommended_strategies"]
    )


def assess_risk(
    volatility: VolatilityAssessment,
    solvency: SolvencyCheck,
    price_vs_target: float,
    quote: dict
) -> Tuple[RiskLevel, DirectionBias, bool]:
    """
    Determine overall risk level and direction bias.

    MCP Tool: assess_risk

    Returns: (risk_level, direction_bias, is_safe_play)
    """
    # Start with volatility as base risk
    risk_score = {"LOW": 1, "MEDIUM": 2, "HIGH": 3, "EXTREME": 4}[volatility.category.value]

    # Adjust for solvency
    if not solvency.is_solvent:
        risk_score += 1

    # Check price extension
    fifty_day_ma = quote.get("fifty_day_ma")
    current_price = quote.get("price")
    if fifty_day_ma and current_price and fifty_day_ma > 0:
        extension = (current_price - fifty_day_ma) / fifty_day_ma * 100
        if abs(extension) > 30:
            risk_score += 1

    # Determine direction bias
    if price_vs_target > 10:
        direction_bias = DirectionBias.BULLISH
    elif price_vs_target < -10:
        direction_bias = DirectionBias.BEARISH
    else:
        direction_bias = DirectionBias.NEUTRAL

    # Determine risk level
    if risk_score <= 1:
        risk_level = RiskLevel.LOW
    elif risk_score <= 2:
        risk_level = RiskLevel.MEDIUM
    elif risk_score <= 3:
        risk_level = RiskLevel.HIGH
    else:
        risk_level = RiskLevel.EXTREME

    # Safe play = solvent + not extreme risk
    is_safe = solvency.is_solvent and risk_level in [RiskLevel.LOW, RiskLevel.MEDIUM]

    return risk_level, direction_bias, is_safe


def _build_recommendation(
    symbol: str,
    risk_level: RiskLevel,
    direction_bias: DirectionBias,
    is_safe: bool,
    volatility: VolatilityAssessment,
    solvency: SolvencyCheck
) -> str:
    """Build human-readable recommendation summary."""
    parts = [f"{symbol}:"]

    if is_safe:
        parts.append("SAFE PLAY -")
    else:
        parts.append("SPECULATIVE -")

    parts.append(f"{risk_level.value} risk,")
    parts.append(f"{direction_bias.value} bias.")

    if volatility.category in [RiskLevel.HIGH, RiskLevel.EXTREME]:
        parts.append(f"High volatility ({volatility.volatility_annualized:.0f}%) favors directional strategies.")
    else:
        parts.append(f"Moderate volatility ({volatility.volatility_annualized:.0f}%) allows income strategies.")

    if not solvency.is_solvent:
        parts.append("WARNING: Negative operating cash flow.")

    return " ".join(parts)

File: apps/api/src/services/analyzer.ts
/**
 * Deep Dive Analyzer Service - Fundamental and technical analysis
 *
 * Checks solvency, calculates volatility, assesses risk level.
 */

import { db, schema } from '@stock-researcher/db';
import { yahooFinance } from './yahoo-finance.js';
import { alphaVantage } from './alpha-vantage.js';
import type {
  AnalysisResponse,
  SolvencyCheck,
  VolatilityAssessment,
  RiskLevel,
  DirectionBias,
} from '@stock-researcher/shared';

interface AnalyzeRequest {
  symbol: string;
  scan_id: number | null;
}

/**
 * Perform full deep dive analysis on a ticker.
 */
async function analyze(request: AnalyzeRequest): Promise<AnalysisResponse> {
  const symbol = request.symbol.toUpperCase();

  // Step 1: Get quote
  const quote = await yahooFinance.getQuote(symbol);
  const currentPrice = quote.price;

  if (!currentPrice) {
    throw new Error(`Could not get price for ${symbol}`);
  }

  // Step 2: Solvency check
  const solvency = await checkSolvency(symbol);

  // Step 3: Volatility
  const volatility = await assessVolatility(symbol);

  // Step 4: Analyst ratings
  const ratings = await yahooFinance.getAnalystRatings(symbol);

  // Step 5: Risk assessment
  const { riskLevel, directionBias, isSafe } = assessRisk(
    volatility,
    solvency,
    ratings.upside_potential_pct ?? 0,
    quote
  );

  // Build recommendation
  const recommendation = buildRecommendation(symbol, riskLevel, directionBias, isSafe, volatility, solvency);

  // Step 6: Persist
  const [analysis] = await db
    .insert(schema.analysisResults)
    .values({
      scanId: request.scan_id,
      symbol,
      createdAt: new Date(),
      currentPrice,
      priceChangePct: quote.change_percent,
      fiftyTwoWeekHigh: quote.fifty_two_week_high,
      fiftyTwoWeekLow: quote.fifty_two_week_low,
      operatingCashFlow: solvency.operating_cash_flow,
      isSolvent: solvency.is_solvent,
      volatility20d: volatility.volatility_annualized,
      volatilityCategory: volatility.category,
      analystRating: ratings.recommendation_key,
      analystTargetMean: ratings.target_mean,
      upsidePotentialPct: ratings.upside_potential_pct,
      riskLevel,
      directionBias,
      isSafePlay: isSafe,
      rawFundamentals: ratings as unknown as Record<string, unknown>,
      rawTechnicals: quote as unknown as Record<string, unknown>,
    })
    .returning();

  // Step 7: Return response
  return {
    analysis_id: analysis.id,
    symbol,
    current_price: currentPrice,
    solvency,
    volatility,
    analyst_rating: ratings.recommendation_key,
    analyst_target_mean: ratings.target_mean,
    upside_potential_pct: ratings.upside_potential_pct,
    risk_level: riskLevel,
    direction_bias: directionBias,
    is_safe_play: isSafe,
    recommendation_summary: recommendation,
  };
}

/**
 * Check solvency via Operating Cash Flow.
 * Uses Alpha Vantage CASH_FLOW endpoint.
 */
async function checkSolvency(symbol: string): Promise<SolvencyCheck> {
  try {
    const cfData = await alphaVantage.getCashFlow(symbol);

    if (cfData.error) {
      return {
        operating_cash_flow: null,
        free_cash_flow: null,
        is_solvent: true,
        notes: 'Unable to verify OCF - assuming solvent (verify manually)',
      };
    }

    const ocf = cfData.operatingCashFlow;
    const fcf = cfData.freeCashFlow;

    return {
      operating_cash_flow: ocf,
      free_cash_flow: fcf,
      is_solvent: ocf !== null && ocf > 0,
      notes: ocf !== null && ocf > 0 ? 'Positive OCF indicates operational solvency' : 'Negative OCF - cash burn',
    };
  } catch {
    return {
      operating_cash_flow: null,
      free_cash_flow: null,
      is_solvent: true,
      notes: 'Unable to verify OCF - assuming solvent (verify manually)',
    };
  }
}

/**
 * Calculate and categorize 20-day volatility.
 */
async function assessVolatility(symbol: string): Promise<VolatilityAssessment> {
  try {
    const volData = await yahooFinance.calculateVolatility(symbol, 20);

    return {
      volatility_20d: volData.daily_volatility,
      volatility_annualized: volData.annualized_volatility,
      category: volData.category,
      recommended_strategies: volData.recommended_strategies,
    };
  } catch {
    return {
      volatility_20d: 1.5,
      volatility_annualized: 30.0,
      category: 'MEDIUM',
      recommended_strategies: ['Unable to calculate - use caution'],
    };
  }
}

/**
 * Determine overall risk level and direction bias.
 */
function assessRisk(
  volatility: VolatilityAssessment,
  solvency: SolvencyCheck,
  priceVsTarget: number,
  quote: { fifty_day_ma: number | null; price: number | null }
): { riskLevel: RiskLevel; directionBias: DirectionBias; isSafe: boolean } {
  // Start with volatility as base risk
  const riskScoreMap: Record<RiskLevel, number> = { LOW: 1, MEDIUM: 2, HIGH: 3, EXTREME: 4 };
  let riskScore = riskScoreMap[volatility.category];

  // Adjust for solvency
  if (!solvency.is_solvent) {
    riskScore += 1;
  }

  // Check price extension
  const fiftyDayMa = quote.fifty_day_ma;
  const currentPrice = quote.price;
  if (fiftyDayMa && currentPrice && fiftyDayMa > 0) {
    const extension = ((currentPrice - fiftyDayMa) / fiftyDayMa) * 100;
    if (Math.abs(extension) > 30) {
      riskScore += 1;
    }
  }

  // Determine direction bias
  let directionBias: DirectionBias;
  if (priceVsTarget > 10) {
    directionBias = 'BULLISH';
  } else if (priceVsTarget < -10) {
    directionBias = 'BEARISH';
  } else {
    directionBias = 'NEUTRAL';
  }

  // Determine risk level
  let riskLevel: RiskLevel;
  if (riskScore <= 1) {
    riskLevel = 'LOW';
  } else if (riskScore <= 2) {
    riskLevel = 'MEDIUM';
  } else if (riskScore <= 3) {
    riskLevel = 'HIGH';
  } else {
    riskLevel = 'EXTREME';
  }

  // Safe play = solvent + not extreme risk
  const isSafe = solvency.is_solvent && (riskLevel === 'LOW' || riskLevel === 'MEDIUM');

  return { riskLevel, directionBias, isSafe };
}

/**
 * Build human-readable recommendation summary.
 */
function buildRecommendation(
  symbol: string,
  riskLevel: RiskLevel,
  directionBias: DirectionBias,
  isSafe: boolean,
  volatility: VolatilityAssessment,
  solvency: SolvencyCheck
): string {
  const parts: string[] = [`${symbol}:`];

  if (isSafe) {
    parts.push('SAFE PLAY -');
  } else {
    parts.push('SPECULATIVE -');
  }

  parts.push(`${riskLevel} risk,`);
  parts.push(`${directionBias} bias.`);

  if (volatility.category === 'HIGH' || volatility.category === 'EXTREME') {
    parts.push(`High volatility (${volatility.volatility_annualized.toFixed(0)}%) favors directional strategies.`);
  } else {
    parts.push(`Moderate volatility (${volatility.volatility_annualized.toFixed(0)}%) allows income strategies.`);
  }

  if (!solvency.is_solvent) {
    parts.push('WARNING: Negative operating cash flow.');
  }

  return parts.join(' ');
}

// Export as module
export const analyzer = {
  analyze,
  checkSolvency,
  assessVolatility,
};
